feat : Add Remaining GraphQL APIs and Refactor SDK Implementation

SkySingh04 opened the pull request:
# PR Update: Litmus Go SDK Refactoring

## SDK Architecture Overview

We have completely refactored the Litmus Go codebase into a proper SDK-like interface with the following design principles:

1. **Interface-based design**: Each feature area has a dedicated interface
2. **Method chaining**: Operations can be chained for better readability (`client.Projects().List()`) 
3. **Fluent API**: The interface is intuitive and self-documenting
4. **Consistent error handling**: All operations return appropriate results and errors

## SDK Structure

- `Client`: Main entry point providing access to all feature areas
- Feature-specific clients:
  - `ProjectClient`: Project management operations
  - `AuthClient`: Authentication operations
  - `EnvironmentClient`: Environment management
  - `ExperimentClient`: Experiment creation and execution
  - `InfrastructureClient`: Infrastructure management
  - `ProbeClient`: Probe operations

## APIs Covered

### Environment Package
- **CreateEnvironment**: Creates a new chaos environment
- **ListChaosEnvironments**: Lists all chaos environments for a project
- **GetChaosEnvironment**: Retrieves a specific environment by ID
- **DeleteEnvironment**: Deletes an environment

### Infrastructure Package
- **GetInfraList**: Retrieves a list of chaos infrastructures
- **ConnectInfra**: Connects/registers a new infrastructure
- **DisconnectInfra**: Disconnects an existing infrastructure
- **GetServerVersion**: Retrieves the server version information

### Probe Package
- **GetProbeRequest**: Retrieves a specific probe by ID
- **ListProbeRequest**: Lists probes with optional type filtering
- **DeleteProbeRequest**: Deletes a probe
- **GetProbeYAMLRequest**: Gets the YAML representation of a probe

### Experiment Package
The experiment client provides methods for managing chaos experiments with a clean interface.

## Usage Examples

```go
// Initialize the client
client, err := sdk.NewClient(sdk.ClientOptions{
    Endpoint: "https://litmus.example.com",
    Username: "admin",
    Password: "password",
})

// Work with projects
projects, err := client.Projects().List()

// Create and run experiments
expConfig := map[string]interface{}{
    "type":      "pod-delete",
    "target":    "deployment/nginx",
    "namespace": "default",
    "duration":  30,
}
newExp, err := client.Experiments().Create("nginx-test", expConfig)
result, err := client.Experiments().Run("experiment-id")

// Manage infrastructure
client.Infrastructure().Connect("infra-id", connectParams)
```

## Testing Approach
- Consistent test structure across all packages following the existing pattern
- Each API has test cases for both successful and error scenarios
- Tests include validation functions to verify response structure and data
- Configurable test environment through environment variables
- Centralized client setup and authentication
- Reusable test helper functions to improve maintainability

## "mongo: no documents in result" Errors

These tests fail because they can't find expected documents in the database:

### Environment API
- **TestGetChaosEnvironment/successful_environment_retrieval** - Cannot find environment record
- **TestDeleteEnvironment/successful_environment_deletion** - Cannot find environment to delete

### Experiment API
- **TestSaveExperiment/successful_experiment_save** - Cannot find infrastructure details
- **TestRunExperiment/successful_experiment_run** - Cannot find experiment to run
- **TestDeleteChaosExperiment/successful_experiment_deletion** - Cannot find experiment to delete
- **TestCreateExperiment/successful_experiment_creation_and_run** - Cannot find infrastructure details

### Infrastructure API
- **TestDisconnectInfra/successful_infrastructure_disconnection** - Cannot find infrastructure to disconnect

### Probe API
- **TestGetProbeRequest/successful_probe_retrieval** - Cannot find probe
- **TestDeleteProbeRequest/successful_probe_deletion** - Cannot find probe to delete

This is because in our tests we were testing the real GraphQL API calls with a real MongoDB database. This will be fixed in future iterations.

SkySingh04 committed:
Update test endpoint and add environment and infrastructure APIs

- Added new APIs for managing environments and infrastructures, including creation, listing, retrieval, and deletion functionalities.
- Introduced GraphQL queries and types for environment and infrastructure operations.
- Enhanced test coverage for the new APIs with comprehensive test cases.

Signed-off-by: Sky Singh <akashsingh2210670@gmail.com>

SkySingh04 committed:
Add SDK components for authentication and project management

- Introduced AuthClient and ProjectClient interfaces for handling authentication and project operations.
- Implemented LitmusClient to manage API interactions, including authentication and project functionalities.
- Added methods for token retrieval, credential management, and project listing, creation, and details retrieval.

Signed-off-by: [Your Name] <[Your Email]>
Signed-off-by: Sky Singh <akashsingh2210670@gmail.com>

SkySingh04 committed:
Refactor client implementation to use new SDK structure

- Updated the client initialization to utilize the new SDK interface, replacing the previous LitmusClient structure with a more modular approach.
- Introduced methods for managing environments, experiments, infrastructure, and probes, enhancing the overall functionality of the SDK.
- Updated tests to reflect changes in client methods and ensure compatibility with the new SDK design.

Signed-off-by: [Your Name] <[Your Email]>
Signed-off-by: Sky Singh <akashsingh2210670@gmail.com>

SkySingh04 committed:
Add complete SDK example for project, environment, experiment, infrastructure, and probe operations

- Introduced a new example file demonstrating the usage of the Litmus SDK.
- Included operations for managing projects, environments, experiments, infrastructure, and probes.
- Implemented error handling and logging for each operation to ensure clarity in execution.

Signed-off-by: [Your Name] <[Your Email]>
Signed-off-by: Sky Singh <akashsingh2210670@gmail.com>

ispeakc0de commented on pkg/sdk/probe.go:
```diff
@@ -0,0 +1,82 @@
+
+// List retrieves all probes
+func (c *probeClient) List() (interface{}, error) {
+	// TODO: Implement when probe API is available
```
You can update these functions now

SkySingh04 replied:
These functions have been updated to call the respective APIs!

SkySingh04 committed:
Enhance ProbeClient interface with projectID parameter and implement basic error handling

- Updated the ProbeClient interface methods to include projectID as a parameter for List, Create, Delete, Update, Get, and Execute operations.
- Implemented error handling for missing credentials, empty project IDs, and probe IDs in respective methods.
- Added functionality to list and delete probes using the new projectID parameter, with appropriate error messages for failed operations.

Signed-off-by: [Your Name] <[Your Email]>
Signed-off-by: Sky Singh <akashsingh2210670@gmail.com>

SkySingh04 committed:
Refactor SDK example and ProbeClient interface for improved probe management

- Updated the CompleteSDKExample function to list probes using a specific probe ID.
- Removed unused methods for creating, updating, and executing probes from the ProbeClient interface.
- Introduced a new method, GetProbeYAML, to retrieve the YAML configuration for a probe, enhancing the interface's functionality.

Signed-off-by: [Your Name] <[Your Email]>
Signed-off-by: Sky Singh <akashsingh2210670@gmail.com>

SkySingh04 committed:
Refactor SDK components for environment, experiment, and infrastructure management

- Removed unused methods from EnvironmentClient and ExperimentClient interfaces, streamlining the SDK.
- Implemented error handling for missing credentials and project IDs in various methods.
- Enhanced List, Create, Delete, and Get operations for environments and experiments with appropriate error messages.
- Updated infrastructure management methods to include connection handling and improved error reporting.

Signed-off-by: [Your Name] <[Your Email]>
Signed-off-by: Sky Singh <akashsingh2210670@gmail.com>

SkySingh04 committed:
Remove outdated comments in Get methods for ExperimentClient and InfrastructureClient to improve code clarity.

Signed-off-by: Sky Singh <akashsingh2210670@gmail.com>
